### ISSUE TITLE
Add support for the new native ConPTY API in Windows 10

### ISSUE THREAD

**Issue by zacps** [jwilm/alacritty#1661]
Information about the API is located here: https://blogs.msdn.microsoft.com/commandline/2018/08/02/windows-command-line-introducing-the-windows-pseudo-console-conpty/

* Add bindings for the new API. Ideally we'd add support for this to winapi-rs, ~~~but I'm not sure how they deal with insider APIs~~~. (https://github.com/retep998/winapi-rs/issues/695)
* Figure out whether we can use the insider SDK on Appveyor. It may have to be feature-gated if we can't.
    * [The API is currently available on Windows 10 v1809, Windows Server 2019](https://docs.microsoft.com/en-us/windows/console/createpseudoconsole#requirements)
    * [Travis supports Windows Server 2016](https://docs.travis-ci.com/user/reference/windows/)
    * [Appveyor supports Windows Server 2012, 2016](https://www.appveyor.com/docs/windows-images-software/)

* Add an implementation of the `EventedReadWrite` (`src/tty/mod.rs`) trait for the ConPTY API.
* Add config option for PTY backend on Windows and runtime switching based on Windows version.

**Comment by davidhewitt** [comment]
I had some fun today hacking around with the new Pseudoconsole APIs on a branch. Definitely not ready to merge, and I need to upstream the patch I made to winapi. If you're interested: https://github.com/davidhewitt/alacritty/tree/consoleapi

It builds and runs (for me on the Win10 October 2018 update), although the pseudoconsole closing & resizing need to be handled yet.

I plan to spend some time refining this further, and can make a PR once it's complete if that's desirable.

**Comment by zacps** [comment]
Great! A PR is definitely desirable.

**Comment by davidhewitt** [comment]
Are you happy if the design is to check for these APIs using `GetProcAddress` and if they can't be found, fall back to WinPty?

Alternative options involve switching implementation based on a feature gate. Easier when building but won't be as nice for end users having to download different binaries depending on their windows version 😄

**Comment by zacps** [comment]
I want winpty built by default with Conpty behind a feature, AND runtime switching based on GetProcAddress.

I'd also like a config option to force one backend (fine to cash it it isn't available).

**Comment by zacps** [comment]
On second thought the config option can be left out now. I thought it could be useful for debugging purposes but it's probably not a big deal.

**Comment by davidhewitt** [comment]
Progress update:
  - I've got that branch working with `GetProcAddress` to load conpty with a fallback
  - I agree that the config option is desirable to force a backend, not yet implemented
 
Overall it seems stable and performant. I haven't benchmarked against Winpty implementation yet. There are sadly a couple of visual glitches that I am yet to investigate when e.g. resizing the window, I'm not sure yet if the fault is bugs in the windows conpty API itself or me/alacritty mishandling it.

I had to modify the `windows.rs` file a nontrivial amount to make the Conpty and Winpty interfaces compatible. In the interest of not diverging too far from master, and also getting experience from others' use, I plan to open a PR soon which merges this, disabled by default, with a config option for those that want to use to opt in to it.

**Comment by zacps** [comment]
Great :smile:. One thing I noticed looking at your branch is that you switched to dynamic dispatch for `conin`/`conout`. I'd probably rather this was done generically unless there are any big issues with that approach.

It'd also be good if you could split the `Evented for AsyncAnon` into a different crate (like font/copypasta) because it's pretty self contained.

Of course these can all wait in favor of getting it working first.

**Comment by davidhewitt** [comment]
Agreed on moving `Evented for AsyncAnon` - I was intending to put it into a crate local to the repo for the moment. Plan to later either delete it or refine further. I'm slightly suspicious that I've missed an existing solution despite extensive searching!

I switched to dynamic dispatch because the `tty::new()` function has to return a single concrete Pty type. So if the winpty/conpty switching is to be done inside it then the type returned from it must abstract across both types of pipe handle.

I did imagine a static solution based on an enum to wrap the possible pipe types, though chose dynamic typing because the enum lookup seemed to me to be comparable but more boilerplate - we'd have been doing a match on the enum with each read/write rather than following a pointer to a vtable.

Can go back static dispatch if you like the enum solution, or will gladly accept any alternative ideas you have :)
